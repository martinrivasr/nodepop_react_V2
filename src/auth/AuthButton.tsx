import { useLocation, useNavigate } from "react-router-dom";
import { useEffect } from "react";
import { useAppDispatch, useAppSelector } from "../store/index.ts";
import { logout } from "../services/api";
import storage from "../utils/storage.ts"
import { setAuthorizationHeader } from "../services/connection.ts";
import { getIslogged } from "../store/selectors.ts";
import { authLogout } from "../store/actions.ts";


export default function AuthButton() {
  const location = useLocation()
  const navigate = useNavigate();
  const accessToken = storage.get("auth");
  const { isLogged, rememberMe } = useAppSelector(getIslogged);
  const dispatch = useAppDispatch()

  console.log("estado de isLogged en authbutton:", isLogged)
  console.log("estado de rememberMe en authbutton:", rememberMe)
  console.log("estado de accessToken en authbutton:", accessToken)
  useEffect(() => {
    if (accessToken) {
      setAuthorizationHeader(accessToken);
    }
  }, [accessToken]);


  const handleLoginClick = () => {
    if (!isLogged && !accessToken){
      navigate("/login");
    } else {
      const to = location.state?.from ?? "/adverts"
      navigate(to, { replace: true}); 
    }
  };


  const handleLogoutClick = async (rememberMe?:boolean) => {
    await logout(rememberMe ?? false)
    dispatch(authLogout())
  };

  return isLogged ? (
    <button className="btn btn-primary  ms-3" onClick={() => handleLogoutClick(rememberMe) }>
      Logout
    </button>
  ) : (
    <button className="btn btn-primary ms-auto" onClick={handleLoginClick}>
      Login
    </button>
  );
}
