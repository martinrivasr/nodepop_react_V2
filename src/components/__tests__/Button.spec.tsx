import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import AuthButton from "../../auth/AuthButton";
import { Provider } from "react-redux";
import configureStore from "../../store";
import { MemoryRouter } from "react-router-dom";
import { createBrowserRouter } from "react-router-dom";


describe("AuthButton Snapshot Test", () => {
  test("shpuld match with the snapshot when the user is not logged", () => {
    
    const router = createBrowserRouter([{ path: "*", element: <AuthButton /> }]);
    const store = configureStore({ auth: { isLogged: false, rememberMe: false } }, router);
    

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <AuthButton />
        </MemoryRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("shpuld match with the snapshot when the user is logged", () => {
    const router = createBrowserRouter([{ path: "*", element: <AuthButton /> }]);
    const store = configureStore({ auth: { isLogged: false, rememberMe: false } }, router);

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <AuthButton />
        </MemoryRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
