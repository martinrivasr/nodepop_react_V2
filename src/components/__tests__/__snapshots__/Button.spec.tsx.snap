// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthButton Snapshot Test > Debe coincidir con el snapshot cuando el usuario ESTÁ logueado 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary  ms-3"
  >
    Logout
  </button>
</DocumentFragment>
`;

exports[`AuthButton Snapshot Test > Debe coincidir con el snapshot cuando el usuario NO está logueado 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary ms-auto"
  >
    Login
  </button>
</DocumentFragment>
`;

exports[`AuthButton Snapshot Test > shpuld match with the snapshot when the user is logged 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary ms-auto"
  >
    Login
  </button>
</DocumentFragment>
`;

exports[`AuthButton Snapshot Test > shpuld match with the snapshot when the user is not logged 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary ms-auto"
  >
    Login
  </button>
</DocumentFragment>
`;

exports[`AuthButton Snapshot Test > shpuld match with the snapshot when the user is not logged 2`] = `
<DocumentFragment>
  <button
    class="btn btn-primary ms-auto"
  >
    Login
  </button>
</DocumentFragment>
`;
